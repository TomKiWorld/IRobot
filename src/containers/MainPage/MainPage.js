import React, { Component, Suspense } from 'react';

import Header from '../../components/Header/Header';
import Robots from '../Robots/Robots';
import Footer from '../../components/Footer/Footer';

const Personal= React.lazy(() => import('../Personal/Personal'));

class MainPage extends Component {
  componentDidMount() {
    this.props.onRequestRobots();
  }

  filterRobots = (robots) => {
    return robots.filter(robot => {
      return robot.name.toLowerCase().includes(this.props.searchField.toLowerCase());
    })
  };

  renderContent = (route) => {    
    switch(route) {
      case 'personal':
        return(
          <Suspense fallback={<div>Loading...</div>}>
            <Personal 
              onNameChange={this.props.onNameChange}
              name={this.props.name}
            />
          </Suspense>
        );
      default:
        return(
          <Robots 
            isPending={this.props.isPending}
            onSearchChange={this.props.onSearchChange}
            filteredRobots={this.filterRobots(this.props.robots)}
          />
        );
    }
  }

  render() {
    const { onSearchChange, onRouteChange, onNameChange, route } = this.props;
    
    return (
      <article className='tc'>
        <Header 
          onRouteChange={onRouteChange}
          resetSearch={onSearchChange}
          resetName={onNameChange}
        />
        {this.renderContent(route)}
        <Footer />
      </article>
    );
  }
}

export default MainPage;
