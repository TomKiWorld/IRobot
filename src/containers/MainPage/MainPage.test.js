import { shallow } from 'enzyme';
import React from 'react';
import MainPage from './MainPage';

describe('MainPage Container', () => {
  const mockProps = {
    onRequestRobots: jest.fn(),
    robots: [],
    route: '',
    searchField: '',
    isPending: false
  };
  const wrapper = shallow(<MainPage { ...mockProps}/>);

  it('Expect to render MainPage Component with robots', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Filter robots correctly', () => {
    expect(wrapper.instance().filterRobots([])).toEqual([]);
  });

  it('Filter robots correctly based on search field return robot', () => {
    const robots = [{
      id: 3,
      name: 'John',
      email: 'john@gmail.com'
    }];
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: robots,
      route: 'list',
      searchField: 'J',
      isPending: false
    };
    const wrapper = shallow(<MainPage { ...mockProps }/>);
    expect(wrapper.instance().filterRobots(mockProps.robots)).toEqual(robots);
  });

  it('Filter robots correctly based on search field return empty', () => {
    const robots = [{
      id: 3,
      name: 'John',
      email: 'john@gmail.com'
    }];
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: robots,
      route: 'list',
      searchField: 'a',
      isPending: false
    };
    const wrapper = shallow(<MainPage { ...mockProps }/>);
    expect(wrapper.instance().filterRobots(mockProps.robots)).toEqual([]);
  });

  it('Expect to render MainPage Component with personal', () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [],
      route: 'personal',
      name: '',
      searchField: '',
      isPending: false
    };
    const wrapper = shallow(<MainPage { ...mockProps }/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('Expect to render MainPage Component with personal and name', () => {
    const mockProps = {
      onRequestRobots: jest.fn(),
      robots: [],
      route: 'personal',
      name: 'tom',
      searchField: '',
      isPending: false
    };
    const wrapper = shallow(<MainPage { ...mockProps }/>);
    expect(wrapper).toMatchSnapshot();
  });
});
