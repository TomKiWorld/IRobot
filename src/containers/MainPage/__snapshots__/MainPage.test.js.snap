// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage Container Expect to render MainPage Component with personal 1`] = `ShallowWrapper {}`;

exports[`MainPage Container Expect to render MainPage Component with personal and name 1`] = `ShallowWrapper {}`;

exports[`MainPage Container Expect to render MainPage Component with robots 1`] = `ShallowWrapper {}`;
